backtracking/word_ladder.py

```diff
@@ -0,0 +1,76 @@
+"""
+Word Ladder is a classic problem in computer science.
+The problem is to transform a start word into an end word
+by changing one letter at a time.
+Each intermediate word must be a valid word from a given list of words.
+The goal is to find a transformation sequence
+from the start word to the end word.
+
+Wikipedia: https://en.wikipedia.org/wiki/Word_ladder
+"""
+
+
+def word_ladder_backtrack(
+    begin_word: str, end_word: str, word_list: list[str]
+) -> list[str]:
+    """
+    Solve the Word Ladder problem using Backtracking and return
+    the list of transformations from begin_word to end_word.
+
+    Parameters:
+    begin_word (str): The word from which the transformation starts.
+    end_word (str): The target word for transformation.
+    word_list (list[str]): The list of valid words for transformation.
+
+    Returns:
+    list[str]: The list of transformations from begin_word to end_word.
+               Returns an empty list if there is no valid transformation.
+
+    Example:
+    >>> word_ladder_backtrack("hit", "cog", ["hot", "dot", "dog", "lot", "log", "cog"])
+    ['hit', 'hot', 'dot', 'lot', 'log', 'cog']
+
+    >>> word_ladder_backtrack("hit", "cog", ["hot", "dot", "dog", "lot", "log"])
+    []
+
+    >>> word_ladder_backtrack("lead", "gold", ["load", "goad", "gold", "lead", "lord"])
+    ['lead', 'lead', 'load', 'goad', 'gold']
+
+    >>> word_ladder_backtrack("game", "code", ["came", "cage", "code", "cade", "gave"])
+    ['game', 'came', 'cade', 'code']
+    """
+
+    # Step 1: Convert the word_list to a set for faster lookup
+    word_set = set(word_list)
+
+    # Step 2: If end_word is not in the word_list, return an empty list
+    if end_word not in word_set:
+        return []
+
+    # Step 3: Backtracking function to find the word ladder
+    def backtrack(current_word, path):
```

Comment:
Please refactor the `backtrack` function to be a separate function. We don't like having inner functions in this repo because they can't be easily unit tested. Please make sure to also add a docstring, unit tests, and type hints to this function.